tests/providers/test_openai.py

```diff
@@ -2,62 +2,160 @@
 
 import pytest
 
-from exchange import Text, ToolUse
-from exchange.providers.openai import OpenAiProvider
-from .conftest import complete, vision, tools
-
-OPENAI_MODEL = os.getenv("OPENAI_MODEL", "gpt-4o-mini")
-
-
-@pytest.mark.vcr()
-def test_openai_complete(default_openai_env):
-    reply_message, reply_usage = complete(OpenAiProvider, OPENAI_MODEL)
-
-    assert reply_message.content == [Text(text="Hello! How can I assist you today?")]
-    assert reply_usage.total_tokens == 27
+from exchange import Text, Message, ToolUse, ToolResult, Tool
+from exchange.providers import OpenAiProvider, AzureProvider
+from exchange.providers.ollama import OllamaProvider, OLLAMA_MODEL
+from .conftest import mark_parametrized
+from ..conftest import read_file
+
+
+# provider configuration for all tests except vision
+providers = [
+    (OllamaProvider, os.getenv("OLLAMA_MODEL", OLLAMA_MODEL)),
+    (OpenAiProvider, os.getenv("OPENAI_MODEL", "gpt-4o-mini")),
+    (AzureProvider, os.getenv("AZURE_MODEL", "gpt-4o-mini")),
+]
+
+# Currently, only OpenAI (not yet Azure) supports vision in its model.
+vision_providers = [t for t in providers if t[0] == OpenAiProvider]
+
+
+@pytest.mark.vcr
+@mark_parametrized(
+    providers,
```

Comment:
Do you think it's reasonable to allow omitting providers when it's the full list?

Reply:
yeah I agree on this completely

Reply:
PTAL!